### ISSUE TITLE
[QUERY] Messages picked up with a long delay using spring-cloud-azure-stream-binder-servicebus

### ISSUE THREAD

**Issue by LauraDamianTNA** [Azure/azure-sdk-for-java#30609]
Hi,
We have an application that is consuming messages from a service bus queue using spring-cloud-azure-stream-binder-servicebus and our code is similar to the one showed here:
https://docs.microsoft.com/en-us/azure/developer/java/spring-framework/configure-spring-cloud-stream-binder-java-app-with-service-bus
(not using manual checkpointing).
We are using openjdk 11.0.16, spring-boot version 2.5.6 and 4.0.0 for spring-cloud-azure-stream-binder-servicebus.
The application is running in Kubernetes, and there are 2 instances of our application that can pick the messages.
The service bus queue has a messgae lock duration of 5 minutes, a 10 minuted duplicate detection window and 20 for the max delivery count.
The scenario we see is similar to this one:
00.00.00 – a message is sent to the service bus queue
00.05.00 – a message is received with a delivery_count = 1 (read from ServiceBusMessageHeaders.DELIVERY_COUNT) and it is processed.
00.09.00 – a message is received with a delivery_count = 0 and it is processed
It’s odd that the message with the delivery_count = 0 is received 4 minutes after the message with delivery_count = 1.
Our hypothesis is that spring-boot gets the message, but does not have an available thread to process it for a time that is bigger than the lock duration we configured for the queue.
As the lock is lost, the second process gets it and starts processing it and then they may compete to process the same entities.
-    Do you have any suggestions on how I can work around this situation?
-    Given the message id, is there something that we can enable and see what happened to the message lock – when we get it, when we lost it?

***Why is this not a Bug or a feature Request?***
Service bus guarantees the delivery only once, so it works as expected from this point of view. I am not sure why the message with the delivery_count = 0 is logged much later than the message with delivery_count = 1 and why I didn't get a lock exception for the first message sooner.

**Setup (please complete the following information if applicable):**
 - OS: Debian GNU/Linux 11
 - Library/Libraries:  
 com.azure.spring:spring-cloud-azure-stream-binder-servicebus-core:4.0.0
 org.springframework.boot:spring-boot-starter-parent:2.5.6
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by joshfree** [comment]
> Our hypothesis is that spring-boot gets the message, but does not have an available thread to process it for a time that is bigger than the lock duration we configured for the queue.
> As the lock is lost, the second process gets it and starts processing it and then they may compete to process the same entities.

Hi @LauraDamianTNA thank you for reaching out to us via this GitHub issue.  Folks from the Azure Spring and the Service Bus SDK team will follow up with you shortly.

/cc @Azure/azsdk-sb-java

**Comment by fbeltrao** [comment]
Additional settings worth to mention:

```plain
spring.cloud.stream.servicebus.bindings.consumer-in-0.consumer.prefetch-count=0
spring.cloud.stream.servicebus.bindings.consumer-in-0.consumer.max-concurrent-calls=9
spring.cloud.stream.servicebus.bindings.consumer-in-0.consumer.max-auto-lock-renew-duration=60M
```

**Comment by yiliuTo** [comment]
@fbeltrao thanks for sharing the detailed error messages, and does the error show up when the delivery count of the message increases and then the message with delivery_count_1 is sending to the client?

**Comment by fbeltrao** [comment]
@yiliuTo a few additional details:

|Attempt|Start|End|Outcome|
|---|---|---|---|
|RetryCount=1|3:08:57.275 PM|3:14:04.503 PM|error|
|RetryCount=0|3:10:49.477 PM|3:14:57.110 PM|error|
|RetryCount=2|3:14:04.508 PM|3:14:46.967 PM|error|
|RetryCount=3|3:14:46.972 PM|3:14:46.973 PM|completed (we give up after 3 tries)|

Single lock duration error was at 3:14:46.973 PM.
Seems like the first attempt (with retryCount=0) didn't even have the lock when it started processing because transition from retry 1 to 2 and 3 are good.

**Comment by yiliuTo** [comment]
Hi @LauraDamianTNA , for the library `spring-cloud-azure-stream-binder-servicebus`, we are using the `ServiceBusProcessorClient` to receive messages from Service Bus. And your issue looks related with [this one](https://github.com/Azure/azure-sdk-for-java/issues/30483), that I suppose it is caused due to the thread pool is exhaused for providing a new thread to consume messages. Thus the processor can receive messages from Service Bus but hangs for acquiring a thread to consume and then the lock duration gets expired. 

As for your questions:
> Q1: Do you have any suggestions on how I can work around this situation?
A: you could try the workaround mentioned in https://github.com/Azure/azure-sdk-for-java/issues/30483 to adjust your vm option of `-Dreactor.schedulers.defaultBoundedElasticSize`. And we can continue discussing if you have more finding after giving a try of that.

> Q2: Given the message id, is there something that we can enable and see what happened to the message lock – when we get it, when we lost it?
A: `spring-cloud-azure-stream-binder-servicebus` provide some message header related with the lock, see [ServiceBusMessageHeaders](https://github.com/Azure/azure-sdk-for-java/blob/da5cc9d6aa3532f8f05f5250f721180f9d010b2c/sdk/spring/spring-messaging-azure-servicebus/src/main/java/com/azure/spring/messaging/servicebus/support/ServiceBusMessageHeaders.java#L41). And for more details, @anuchandy  could youe help to share how we can get the lock processing details via `ServiceBusProcessorClient`? Thanks

**Comment by liukun-msft** [comment]
Hi @yiliuTo  
The issue may not be the same as https://github.com/Azure/azure-sdk-for-java/issues/30483 but may be related to https://github.com/Azure/azure-sdk-for-java/issues/30530.

Hi @LauraDamianTNA and @fbeltrao
Except this message, does the application process for other messages? base on current content, it seems the process received message(delivery_count=0) but haven't processed it until lock timeout at 3:08:57.259, so the message(delivery_count=1) received again in other thread at 3:08:57.275. But what is error for delivery_count=1 and delivery_count=2, each end at 3:14:04.503 and 3:14:46.967?

Can you also provide more detailed logs (with thread name and all errors/warnings during that time) to help us understand your scenarios more?

**Comment by fbeltrao** [comment]
@yiliuTo the prefetching of an additional message would explain what is happening.

The out of order message ID is `4987585e-70a8-40d2-846a-0953de65a903` which was produced at `3:03:57.094`.

We have two instances of our application processing message from the queue. This the processing start of message `4987585e-70a8-40d2-846a-0953de65a903`:

|Instance|Retry-Count|Start|
|---|---|---|
|584bc9fb6b-6sljv|0|3:10:49.478|
|584bc9fb6b-xnsw2|1| 3:08:57.276|
|584bc9fb6b-xnsw2|2|3:14:04.508|
|584bc9fb6b-6sljv|3|3:14:46.972|

Around that time, the instance `584bc9fb6b-6sljv` which had the delayed processing start received the following messages:

|Message ID|Retry-Count|Time|
|---|---|---|
|e4e542cd-6dec-4190-b15f-8a4653c0e8bd|0|3:03:54.592|
|4987585e-70a8-40d2-846a-0953de65a903|0|3:10:49.478|
|4987585e-70a8-40d2-846a-0953de65a903|3|3:14:46.972|

Why do retries 1,2 and 3 fail?
As the retry=0 is running in parallel they end up fighting for resources. We abandon the message once we try 3 times.

For additional context, this application is consuming messages from 2 queues from same service bus. Not sure if that information is helpful.

**Comment by liukun-msft** [comment]
Thanks @fbeltrao for providing these details. This is really helpful.

To help us check the root cause, can you follow https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/servicebus/azure-messaging-servicebus#enable-client-logging to provide the debug level log with period of at least +/- 10 minutes from when the issue occurred?

**Comment by liukun-msft** [comment]
Sorry, currently the enable client logging wiki link is not valid, please directly use https://docs.microsoft.com/en-us/azure/developer/java/sdk/logging-overview#next-steps and choose one logging framework and configure to DEBUG level log. 

Please also enable DEBUG level for AppInsights followed by: https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config#auto-collected-logging

**Comment by liukun-msft** [comment]
We found the root cause was that the processing thread is blocked by the small default thread pool size, so we asked user to add the VM option "-Dreactor.schedulers.defaultBoundedElasticSize=50" to increase the pool size. Considering that things have worked well since they changed it and they have had no more questions, this issue is closed.

If you have more questions, please re-open this one or create a new issue.
